File: 7_graph/拓扑排序/1203. 项目管理.ts
// group[i] 表示第 i 个项目所属的小组，如果这个项目目前无人接手，那么 group[i] 就等于 -1
// 同一小组的项目，排序后在列表中彼此相邻。 -1 表示不属于任一个组
// 项目之间存在一定的依赖关系，我们用一个列表 beforeItems 来表示，其中 beforeItems[i] 表示在进行第 i 个项目前（位于第 i 个项目左侧）应该完成的所有项目。
// n 个项目和  m 个小组
function sortItems(n: number, m: number, group: number[], beforeItems: number[][]): number[] {}

console.log(sortItems(8, 2, [-1, -1, 1, 0, 0, 1, 0, -1], [[], [6], [5], [6], [3, 6], [], [], []]))
// 输出：[6,3,4,1,5,2,0,7]

// 太难了 不做

File: 7_graph/拓扑排序/1203. 项目管理.py
# 利用扩展的拓扑排序一次性求出结果
# https://leetcode.cn/problems/sort-items-by-groups-respecting-dependencies/solutions/467233/li-yong-kuo-zhan-de-tuo-bu-pai-xu-yi-ci-xing-qiu-c/

from typing import List


class Solution:
    def sortItems(
        self, n: int, m: int, group: List[int], beforeItems: List[List[int]]
    ) -> List[int]:
        dep = [0] * n
        group_dep = [0] * m
        other_queue = []
        group_queue = [[] for _ in range(m)]
        queue_queue = []
        dependedBy = [[] for _ in range(n)]
        for i, l in enumerate(beforeItems):
            dep[i] = len(l)
            gi = group[i]
            for j in l:
                dependedBy[j].append(i)
                if gi != -1 and group[j] != gi:
                    group_dep[gi] += 1
            if not l:
                (group_queue[gi] if gi != -1 else other_queue).append(i)
        for i, gd in enumerate(group_dep):
            if not gd:
                queue_queue.append(group_queue[i])
        result = []
        while True:
            if queue_queue:
                current_queue = queue_queue.pop()
            elif other_queue:
                current_queue = other_queue
            else:
                if len(result) < n:
                    return []
                else:
                    return result
            while current_queue:
                p = current_queue.pop()
                result.append(p)
                gp = group[p]
                for i in dependedBy[p]:
                    gi = group[i]
                    dep[i] -= 1
                    if not dep[i]:
                        (group_queue[gi] if gi != -1 else other_queue).append(i)
                    if gi != -1 and gi != gp:
                        group_dep[gi] -= 1
                        if not group_dep[gi]:
                            queue_queue.append(group_queue[gi])
